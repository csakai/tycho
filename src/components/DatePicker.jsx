import React from 'react';
import Datetime from 'react-datetime';
import moment from 'moment';

export default class DatePicker extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = {};
  }
  
  componentWillReceiveProps = () => {
    const {time} = this.props;
    const timeInstance = moment(time * 1000);
    const realTime = timeInstance.toDate();
    const uxTime = this.getUXTime(timeInstance);
    
    if (!this.isOpen) {
      this.setState({uxTime, realTime});
    }
  }

  getUXTime = (timeInstance) => {
    return timeInstance
      .format('MMM DD, YYYY h:mm a') // TODO: move to const
      .replace(/ /g, '\u00A0');
  }
  
  shouldComponentUpdate = () => {
    return !this.isOpen;
  }
  
  hidePicker = () => {
    this.isOpen = false;
  }
  
  showPicker = () => {
    const {picker} = this.refs;
    
    if (picker) {
      this.isOpen = true;
      picker.openCalendar();
    }
  }
  
  changeTime = (timeInstance) => {
    const time = timeInstance.unix();
    this.props.onUpdate(time);
  }
  
  render() {
    return (
      <div className="clock">
        <span
          className="clock__display"
          onClick={this.showPicker}>
          {this.state.uxTime}
        </span>
        <Datetime
          value={this.state.realTime}
          ref="picker"
          className="clock__picker"
          onBlur={this.hidePicker}
          onChange={this.changeTime}
          closeOnSelect={true}
          inputProps={{
            className: 'clock__input'
          }}
        />
      </div>
    );
  }
}
