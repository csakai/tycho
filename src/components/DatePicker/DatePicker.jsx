import React from 'react';

export default class DatePicker extends React.Component {
  render() {
    return (
      <div className="clock">
        <span
          className="clock__display"
          onClick={this.props.onClick}>
          {this.props.uxTime}
        </span>
        {this.props.children}
      </div>
    );
  }
}
