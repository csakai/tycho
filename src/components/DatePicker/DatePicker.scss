
$datepicker-fg-color: #000;
$datepicker-bg-color: #fff;

.clock {
  position: absolute;
  top: 0;
  right: 0;
  min-width: 250px;
  padding: 1em;
  text-align: right;
  
  &__display {
    font-family: Georgia, serif;
    font-size: 2em;
    color: $datepicker-fg-color;
    transition: text-shadow 0.5s;
    cursor: pointer;
    
    &:hover {
      text-shadow: 1px 1px 10px $datepicker-fg-color;
    }
  }
  
  &__picker {
    display: flex;
    justify-content: flex-end;
    color: $datepicker-fg-color;
    font-family: Georgia, serif;
    font-size: 0.9em;
    margin-top: 0.5em;
    
    .rdtPicker {
      background: $datepicker-bg-color;
      
      th, td {
        &:hover {
          background: #1e1e1e !important;
        }
      }
      
      .rdtPrev, .rdtNext {
        border-color: $datepicker-bg-color;
      }
      
      .rdtCounter .rdtBtn:hover {
        background: $datepicker-bg-color;
      }
    }
  }
  
  &__input {
    display: none;
  }
}
