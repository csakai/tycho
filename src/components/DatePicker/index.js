import DatePicker from './DatePicker';
import './DatePicker.css';

export default DatePicker;
