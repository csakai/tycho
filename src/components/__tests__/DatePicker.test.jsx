import React from 'react';
import moment from 'moment';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import DatePicker from '../DatePicker';

describe('DatePicker Component', () => {
  let component, datePicker;

  beforeEach(() => {
    component = shallow(<DatePicker 
      time={1}
      onUpdate={jest.fn()}
    />);

    datePicker = component.instance();
  });

  describe('componentWillReceiveProps()', () => {
    it('should set uxTime and realTime state values if the calendar is not open', () => {
      const uxTime = '11/11/1111';
      const realTime = 123456;

      component.setState({uxTime, realTime});
      component.setProps({time: 54321});

      expect(component.state('uxTime')).not.toEqual(uxTime);
      expect(component.state('realTime')).not.toEqual(realTime);
    });

    it('should not change uxTime or realTime state values if the calendar is already open', () => {
      const uxTime = '11/11/1111';
      const realTime = 123456;

      datePicker.isOpen = true;
      component.setState({uxTime, realTime});
      component.setProps({time: 54321});
      
      expect(component.state('uxTime')).toEqual(uxTime);
      expect(component.state('realTime')).toEqual(realTime);
    });
  });

  describe('getUXTime()', () => {
    it('should return a string', () => {
      const result = datePicker.getUXTime(moment());

      expect(typeof result).toBe('string');
    });
  });

  describe('shouldComponentUpdate()', () => {
    it('should return the inverse of isOpen', () => {
      const result = datePicker.shouldComponentUpdate();
      
      expect(typeof result).toBe('boolean');
      expect(result).toEqual(!datePicker.isOpen);
    });
  });

  describe('hidePicker()', () => {
    it('should set isOpen to false', () => {
      datePicker.isOpen = true;
      datePicker.hidePicker();

      expect(datePicker.isOpen).toEqual(false);
    });
  });

  describe('showPicker()', () => {
    let picker, spy;

    beforeEach(() => {
      picker = {openCalendar: jest.fn()};
      spy = jest.spyOn(picker, 'openCalendar');
    });

    it('should open the calendar if its ref is defined', () => {
      datePicker.refs = {picker};
      datePicker.showPicker();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should not open the calendar if its ref is undefined', () => {
      datePicker.refs = {};
      datePicker.showPicker();

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('changeTime()', () => {
    it('should call the onUpdate prop method with the given unix time', () => {
      datePicker.props = {onUpdate: jest.fn()};

      const time = moment();
      const spy = jest.spyOn(datePicker.props, 'onUpdate');

      datePicker.changeTime(time);

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(time.unix());
    });
  });

  describe('render()', () => {
    it('should render the DatePicker component successfully', () => {
      component.setState({realTime: 123456});

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
