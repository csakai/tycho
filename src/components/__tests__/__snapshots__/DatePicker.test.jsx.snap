// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker Component render() should render the DatePicker component successfully 1`] = `
<div
  className="clock"
>
  <span
    className="clock__display"
    onClick={[Function]}
  />
  <Component
    className="clock__picker"
    closeOnSelect={true}
    closeOnTab={true}
    dateFormat={true}
    defaultValue=""
    input={true}
    inputProps={
      Object {
        "className": "clock__input",
      }
    }
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onViewModeChange={[Function]}
    strictParsing={true}
    timeConstraints={Object {}}
    timeFormat={true}
    utc={false}
    value={123456}
  />
</div>
`;
